### PR TITLE
[SPARK-44931] Fix JSON Serialization of Spark Connect protos for Event Listener

### PR DESCRIPTION
**Pull request by grundprinzip**
### What changes were proposed in this pull request?
This patch fixes an issue with the serialization of the Spark Connect proto messages during the event notifications. 

Without this patch upon receiving Spark Connect events, the Spark log will be full of the following messages:

```
23/08/23 08:55:52 WARN DBCEventLoggingListener: Error in writing the event to log
com.fasterxml.jackson.databind.exc.InvalidDefinitionException: Direct self-reference leading to cycle (through reference chain: org.apache.spark.sql.connect.service.SparkListenerConnectOperationStarted["planRequest"]->org.apache.spark.connect.proto.ExecutePlanRequest["unknownFields"]->com.google.protobuf.UnknownFieldSet["defaultInstanceForType"])
	at com.fasterxml.jackson.databind.exc.InvalidDefinitionException.from(InvalidDefinitionException.java:77)
	at com.fasterxml.jackson.databind.SerializerProvider.reportBadDefinition(SerializerProvider.java:1308)
	at com.fasterxml.jackson.databind.ser.BeanPropertyWriter._handleSelfReference(BeanPropertyWriter.java:948)
	at com.fasterxml.jackson.databind.ser.BeanPropertyWriter.serializeAsField(BeanPropertyWriter.java:726)
	at com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeFields(BeanSerializerBase.java:772)
	at com.fasterxml.jackson.databind.ser.BeanSerializer.serialize(BeanSerializer.java:178)
```

The reason is that the JSON serializer cannot properly deal with recursive nested structures and for that reason, a custom JSON serializer is added that uses the built-in protobuf serialization.

### Why are the changes needed?
Stability / Compatibility

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Existing UT

### CODE REVIEW

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/ExecuteEventsManager.scala [hunk 1]
```diff
@@ -27,6 +31,14 @@ import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
 import org.apache.spark.sql.connect.common.ProtoUtils
 import org.apache.spark.util.{Clock, Utils}
 
+private[sql] class ProtobufSerializer extends JsonSerializer[Message] {
```
Comment (LuciferYang) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/ExecuteEventsManager.scala:34]: Can [sql] be removed?

**Comment by LuciferYang** [comment]
@grundprinzip Sorry, I didn't find a way to validate this patch. Is there any chance you could add a new test case or describe how to verify it manually?

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/ExecuteEventsManager.scala (hunk 1 above)
Comment (grundprinzip) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/ExecuteEventsManager.scala:34]: I can remove the `[sql]`

**Comment by jdesjean** [comment]
I had a [separate PR](https://github.com/apache/spark/pull/42550) that ignored the field from serde. I'm open to either approach. The other PR has some additional testing that could be used here.

**Pull request closed without merging**